cub/device/dispatch/dispatch_segmented_sort.cuh

```diff
@@ -1390,130 +1611,60 @@ private:
               d_keys.Current(),
-              GetFinalOutput(LargeSegmentPolicyT::RADIX_BITS, d_keys),
+              GetFinalOutput<KeyT>(LargeSegmentPolicyT::RADIX_BITS, d_keys),
               d_keys_double_buffer,
               d_values.Current(),
-              GetFinalOutput(LargeSegmentPolicyT::RADIX_BITS, d_values),
+              GetFinalOutput<ValueT>(LargeSegmentPolicyT::RADIX_BITS, d_values),
               d_values_double_buffer,
               d_begin_offsets,
-              d_end_offsets);
-
-      // Check for failure to launch
-      if (CubDebug(error = cudaPeekAtLastError()))
-      {
-        return error;
-      }
-
-      // Sync the stream if specified to flush runtime errors
-      if (debug_synchronous)
-      {
-        if (CubDebug(error = SyncStream(stream)))
-        {
-          return error;
-        }
-      }
-    }
-
-    const unsigned int small_segments  = h_group_sizes[1];
-    const unsigned int medium_segments =
-      static_cast<unsigned int>(num_segments) -
-      (large_segments + small_segments);
-
-    const unsigned int small_blocks =
-      DivideAndRoundUp(small_segments,
-                       SmallAndMediumPolicyT::SEGMENTS_PER_SMALL_BLOCK);
-
-    const unsigned int medium_blocks =
-      DivideAndRoundUp(medium_segments,
-                       SmallAndMediumPolicyT::SEGMENTS_PER_MEDIUM_BLOCK);
-
-    const unsigned int small_and_medium_blocks_in_grid = small_blocks +
-                                                         medium_blocks;
-
-    if (small_and_medium_blocks_in_grid)
-    {
-      if (debug_synchronous)
-      {
-        _CubLog("Invoking "
-                "DeviceSegmentedSortKernelSmall<<<%d, %d, 0, %lld>>>()\n",
-                static_cast<int>(small_and_medium_blocks_in_grid),
-                SmallAndMediumPolicyT::BLOCK_THREADS,
-                (long long)stream);
-      }
-
-      THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron(
-        small_and_medium_blocks_in_grid,
-        SmallAndMediumPolicyT::BLOCK_THREADS,
-        0,
-        stream)
-        .doit(small_kernel,
-              small_segments,
-              medium_segments,
-              medium_blocks,
+              d_end_offsets,
+              group_sizes.get(),
+              large_and_medium_segments_indices.get(),
               small_segments_indices.get(),
-              large_and_medium_segments_indices.get() + num_segments -
-                medium_segments,
-              d_keys.Current(),
-              GetFinalOutput(LargeSegmentPolicyT::RADIX_BITS, d_keys),
-              d_values.Current(),
-              GetFinalOutput(LargeSegmentPolicyT::RADIX_BITS, d_values),
-              d_begin_offsets,
-              d_end_offsets);
+              debug_synchronous);
 
-      // Check for failure to launch
       if (CubDebug(error = cudaPeekAtLastError()))
       {
         return error;
       }
-
-      // Sync the stream if specified to flush runtime errors
-      if (debug_synchronous)
```

Comment:
Let's leave the device-side `debug_synchronous` syncs in for now (also add them in the continuation handler if needed).